File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/buy_stock/P309_BestTimeToBuyAndSellStockWithCooldown.java
package com.dsa.leetcode.labuladong.buy_stock;

public class P309_BestTimeToBuyAndSellStockWithCooldown {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        return recursive(prices, 0, 0);
    }

    private int recursive(int[] prices, int i, int status) {
        if (i >= prices.length) {
            return 0;
        }
        // can buy
        if (status == 0) {
            // buy, you can't buy next day
            int p1 = -prices[i] + recursive(prices, i + 1, 1);
            // not buy
            int p2 = recursive(prices, i + 1, status);
            return Math.max(p1, p2);
        } else {
            // can sell
            int p1 = prices[i] + recursive(prices, i + 2, 0);
            // not sell
            int p2 = recursive(prices, i + 1, status);
            return Math.max(p1, p2);
        }
    }

    public static void main(String[] args) {

    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/dynamic_programming/buy_stock/P309_BestTimeToBuyAndSellStockWithCooldown.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        return self.dp(prices)

    def recursive(self, prices, i, status):
        if i >= len(prices):
            return 0
        if status == 0:
            # buy
            p1 = -prices[i] + self.recursive(prices, i + 1, 1)
            # not buy
            p2 = self.recursive(prices, i + 1, status)
            return max(p1, p2)
        else:
            # sell
            p1 = prices[i] + self.recursive(prices, i + 2, 0)
            # not sell
            p2 = self.recursive(prices, i + 1, status)
            return max(p1, p2)

    def dp(self, prices):
        n = len(prices) + 2
        m = 2
        dp = [[0 for i in range(m)] for i in range(n)]
        for i in range(n - 3, -1, -1):
            for status in range(m):
                if status == 0:
                    p1 = -prices[i] + dp[i + 1][1]
                    p2 = dp[i + 1][status]
                    dp[i][status] = max(p1, p2)
                else:
                    p1 = prices[i] + dp[i + 2][0]
                    p2 = dp[i + 1][status]
                    dp[i][status] = max(p1, p2)
        return dp[0][0]


if __name__ == '__main__':
    prices = [1, 2, 3, 0, 2]
    solu = Solution()
    print(solu.maxProfit(prices))
